hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1439,6 +1440,238 @@ private class Emit(
         }
         emitter.emit(outputPType)
 
+      case x@NDArrayQR(nd, mode) =>
+        val ndt = emit(nd)
+        val ndAddress = mb.newField[Long]
+        val ndPType = nd.pType.asInstanceOf[PNDArray]
+
+        val shapeAddress = ndPType.shape.load(region, ndAddress)
+        val shapeTuple = new CodePTuple(ndPType.shape.pType, region, shapeAddress)
+        val shapeArray = (0 until ndPType.shape.pType.nFields).map(shapeTuple[Long](_)).toArray
+
+        val sizeQueryAddress = mb.newLocal[Long]
+
+        val M = shapeArray(0)
+        val N = shapeArray(1)
+        val K = (M < N).mux(M, N)
+        val LDA = M // Possible stride tricks could change this in the future.
+        val LWORK = Region.loadDouble(sizeQueryAddress).toI
+
+        val dataAddress = ndPType.data.load(region, ndAddress)
+
+        val tauPType = PArray(PFloat64Required, true)
+        val tauAddress = mb.newField[Long]
+        val workAddress = mb.newField[Long]
+        val aAddressDGEQRF = mb.newField[Long] // Should be column major
+        val rDataAddress = mb.newField[Long]
+        val aNumElements = ndPType.numElements(shapeArray, mb)
+
+        val infoDGEQRFResult = mb.newLocal[Int]
+        val infoDGEQRFErrorTest = (infoDGEQRFResult < 0).orEmpty(Code._fatal(const("LAPACK error DGEQRF. Error code = ").concat(infoDGEQRFResult.toS)))
+
+        val alwaysNeeded = Code(
+          ndAddress := ndt.value[Long],
+
+          // Make some space for the column major form (which means copying the input)
+          aAddressDGEQRF := ndPType.data.pType.allocate(region, aNumElements.toI),
+          ndPType.data.pType.stagedInitialize(aAddressDGEQRF, aNumElements.toI),
+          ndPType.copyRowMajorToColumnMajor(dataAddress, aAddressDGEQRF, M, N, mb),
+
+          tauAddress := tauPType.allocate(region, K.toI),
+          tauPType.stagedInitialize(tauAddress, K.toI),
+
+          sizeQueryAddress := ndPType.data.pType.allocate(region, 1), // One double, just to hold size query results
+
+          infoDGEQRFResult := Code.invokeScalaObject[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+            M.toI,
+            N.toI,
+            ndPType.data.pType.elementOffset(aAddressDGEQRF, aNumElements.toI, 0),
+            LDA.toI,
+            tauPType.elementOffset(tauAddress, K.toI, 0),
+            sizeQueryAddress,
+            -1
+          ),
+          infoDGEQRFErrorTest,
```

Comment:
I think we'll thank past ourselves if we include which of the two dgeqrf calls triggered the info failure message

Reply:
Is there any more debugging information we can evince from LAPACK before we raise an exception? Is the error code it?

Reply:
As far as I can tell, it's all in the error code: https://www.netlib.org/lapack/lug/node136.html